File: apps/user-app-ios/Sources/Core/Models/Game.swift
import Foundation

// MARK: - Game Model

struct Game: Identifiable, Codable, Sendable {
    let id: Int
    let name: String
    let description: String?
    let clubId: Int
    let clubName: String
    let clubAddress: String?
    let courses: [GameCourse]
    let durationMinutes: Int
    let maxPlayers: Int
    let basePrice: Int
    let isActive: Bool
    let createdAt: Date?
    let timeSlots: [GameTimeSlot]?

    enum CodingKeys: String, CodingKey {
        case id, name, description
        case clubId = "club_id"
        case clubName = "club_name"
        case clubAddress = "club_address"
        case courses
        case durationMinutes = "duration_minutes"
        case maxPlayers = "max_players"
        case basePrice = "base_price"
        case isActive = "is_active"
        case createdAt = "created_at"
        case timeSlots = "time_slots"
    }

    var durationText: String {
        if durationMinutes >= 60 {
            let hours = durationMinutes / 60
            let mins = durationMinutes % 60
            if mins > 0 {
                return "\(hours)시간 \(mins)분"
            }
            return "\(hours)시간"
        }
        return "\(durationMinutes)분"
    }

    var courseNames: String {
        courses.map { $0.name }.joined(separator: ", ")
    }

    var priceRange: (min: Int, max: Int)? {
        guard let slots = timeSlots, !slots.isEmpty else { return nil }
        let prices = slots.map { $0.price }
        return (prices.min() ?? basePrice, prices.max() ?? basePrice)
    }
}

// MARK: - Game Course

struct GameCourse: Identifiable, Codable, Sendable {
    let id: Int
    let name: String
    let holeCount: Int?

    enum CodingKeys: String, CodingKey {
        case id, name
        case holeCount = "hole_count"
    }
}

// MARK: - Game Time Slot

struct GameTimeSlot: Identifiable, Codable, Sendable {
    let id: Int
    let gameId: Int
    let startTime: String
    let endTime: String
    let price: Int
    let maxCapacity: Int
    let availableSlots: Int
    let isPremium: Bool
    let dayOfWeek: Int?

    enum CodingKeys: String, CodingKey {
        case id
        case gameId = "game_id"
        case startTime = "start_time"
        case endTime = "end_time"
        case price
        case maxCapacity = "max_capacity"
        case availableSlots = "available_slots"
        case isPremium = "is_premium"
        case dayOfWeek = "day_of_week"
    }

    var stringId: String { String(id) }

    var availabilityStatus: AvailabilityStatus {
        if availableSlots == 0 {
            return .soldOut
        } else if availableSlots <= 2 {
            return .almostFull
        } else if availableSlots <= 4 {
            return .limited
        } else {
            return .available
        }
    }
}

enum AvailabilityStatus {
    case available
    case limited
    case almostFull
    case soldOut

    var color: String {
        switch self {
        case .available: return "green"
        case .limited: return "yellow"
        case .almostFull: return "red"
        case .soldOut: return "gray"
        }
    }
}

// MARK: - Game Search Parameters

struct GameSearchParams: Sendable {
    var query: String?
    var date: Date?
    var timeOfDay: TimeOfDay?
    var minPrice: Int?
    var maxPrice: Int?
    var minPlayers: Int?
    var sortBy: SortOption
    var sortOrder: SortOrder
    var page: Int
    var limit: Int

    init(
        query: String? = nil,
        date: Date? = nil,
        timeOfDay: TimeOfDay? = nil,
        minPrice: Int? = nil,
        maxPrice: Int? = nil,
        minPlayers: Int? = nil,
        sortBy: SortOption = .price,
        sortOrder: SortOrder = .asc,
        page: Int = 1,
        limit: Int = 20
    ) {
        self.query = query
        self.date = date
        self.timeOfDay = timeOfDay
        self.minPrice = minPrice
        self.maxPrice = maxPrice
        self.minPlayers = minPlayers
        self.sortBy = sortBy
        self.sortOrder = sortOrder
        self.page = page
        self.limit = limit
    }

    enum TimeOfDay: String, CaseIterable, Sendable {
        case all = "전체"
        case morning = "오전"
        case afternoon = "오후"

        var timeRange: (start: String, end: String)? {
            switch self {
            case .all: return nil
            case .morning: return ("06:00", "11:59")
            case .afternoon: return ("12:00", "18:00")
            }
        }
    }

    enum SortOption: String, CaseIterable, Sendable {
        case price = "가격순"
        case name = "이름순"
        case createdAt = "최신순"
    }

    enum SortOrder: String, CaseIterable, Sendable {
        case asc = "오름차순"
        case desc = "내림차순"
    }
}

// MARK: - Game Search Response

struct GameSearchResponse: Codable, Sendable {
    let data: [Game]
    let total: Int
    let page: Int
    let limit: Int
    let totalPages: Int

    enum CodingKeys: String, CodingKey {
        case data, total, page, limit
        case totalPages = "total_pages"
    }
}

File: apps/user-app-android/app/src/main/java/com/parkgolf/app/domain/model/Game.kt
package com.parkgolf.app.domain.model

data class Game(
    val id: Int,
    val name: String,
    val code: String? = null,
    val description: String? = null,
    val clubId: Int,
    val clubName: String,
    val club: GameClub? = null,
    val frontNineCourse: GameCourse? = null,
    val backNineCourse: GameCourse? = null,
    val totalHoles: Int? = null,
    val estimatedDuration: Int,
    val maxPlayers: Int,
    val basePrice: Int,
    val pricePerPerson: Int? = null,
    val weekendPrice: Int? = null,
    val isActive: Boolean = true,
    val timeSlots: List<GameTimeSlot>? = null
) {
    val durationText: String
        get() = if (estimatedDuration >= 60) {
            val hours = estimatedDuration / 60
            val mins = estimatedDuration % 60
            if (mins > 0) "${hours}시간 ${mins}분" else "${hours}시간"
        } else "${estimatedDuration}분"

    val courseNames: String
        get() = listOfNotNull(frontNineCourse?.name, backNineCourse?.name)
            .joinToString(", ")

    val priceText: String
        get() = "${String.format("%,d", basePrice)}원"

    val priceRange: PriceRange?
        get() {
            val prices = timeSlots?.map { it.price } ?: return null
            if (prices.isEmpty()) return null
            return PriceRange(prices.min(), prices.max())
        }
}

data class GameClub(
    val id: Int,
    val name: String,
    val address: String? = null,
    val phoneNumber: String? = null,
    val description: String? = null,
    val imageUrl: String? = null,
    val latitude: Double? = null,
    val longitude: Double? = null
)

data class GameCourse(
    val id: Int,
    val name: String,
    val holes: Int = 9,
    val par: Int? = null,
    val description: String? = null
)

data class GameTimeSlot(
    val id: Int,
    val gameId: Int,
    val startTime: String,
    val endTime: String,
    val availablePlayers: Int,
    val maxPlayers: Int,
    val price: Int,
    val isPremium: Boolean = false,
    val isAvailable: Boolean = true
) {
    val priceText: String
        get() = "${String.format("%,d", price)}원"

    val priceShortText: String
        get() = "₩${price / 1000}k"

    val timeText: String
        get() = "$startTime - $endTime"

    val availabilityText: String
        get() = "$availablePlayers/${maxPlayers}명"

    val availabilityStatus: AvailabilityStatus
        get() = when {
            availablePlayers == 0 -> AvailabilityStatus.SOLD_OUT
            availablePlayers <= 2 -> AvailabilityStatus.ALMOST_FULL
            availablePlayers <= maxPlayers / 2 -> AvailabilityStatus.LIMITED
            else -> AvailabilityStatus.AVAILABLE
        }
}

enum class AvailabilityStatus {
    AVAILABLE,
    LIMITED,
    ALMOST_FULL,
    SOLD_OUT
}

data class PriceRange(
    val min: Int,
    val max: Int
)

data class GameSearchParams(
    val search: String? = null,
    val date: String? = null,
    val timeOfDay: TimeOfDay? = null,
    val minPrice: Int? = null,
    val maxPrice: Int? = null,
    val minPlayers: Int? = null,
    val sortBy: SortOption = SortOption.PRICE,
    val sortOrder: SortOrder = SortOrder.ASC,
    val page: Int = 1,
    val limit: Int = 20
)

enum class TimeOfDay(val label: String, val startTime: String?, val endTime: String?) {
    ALL("전체", null, null),
    MORNING("오전", "06:00", "12:00"),
    AFTERNOON("오후", "12:00", "18:00"),
    EVENING("저녁", "18:00", "22:00")
}

enum class SortOption(val label: String, val value: String) {
    PRICE("가격순", "price"),
    TIME("시간순", "time"),
    AVAILABILITY("잔여석순", "availability")
}

enum class SortOrder(val label: String, val value: String) {
    ASC("낮은순", "asc"),
    DESC("높은순", "desc")
}
